src/static-template.js

```diff
@@ -95,51 +93,58 @@ export function cachedHtmlFor(nodeOrDoc) {
 }
 
 
-/**
- * @param {!./service/ampdoc-impl.AmpDoc} ampdoc
- * @return {{
- *  htmlInternal: !HtmlLiteralTagDef,
- *  cachedHtmlInternal: !HtmlLiteralTagDef,
- * }}
- */
-export function buildHtmlService(ampdoc) {
-  const container = ampdoc.getRootNode().createElement('div');
-
-  /** @type {!TempCache<!Element>|undefined} */
-  let cache;
-
-  /**
-   * @param {!Array<string>} strings
-   * @return {!Element}
-   */
-  const cachedHtmlInternal = strings => {
-    cache = (cache || new TempCache(ampdoc.win, FLUSH_CACHE_AFTER_MACRO_TASKS));
-    const key = strings[0];
-    const seed = cache.has(key) ?
-      cache.get(key) :
-      cache.put(key, htmlInternal(strings));
-    return seed.cloneNode(/* deep */ true);
-  };
-
-  /**
-   * @param {!Array<string>} strings
-   * @return {!Element}
-   */
-  const htmlInternal = strings => {
-    devAssert(strings.length === 1, 'Improper html template tag usage.');
-    container./*OK*/innerHTML = strings[0];
-
-    const el = container.firstElementChild;
-    devAssert(el, 'No elements in template');
-    devAssert(!el.nextElementSibling, 'Too many root elements in template');
-
-    // Clear to free memory.
-    container.removeChild(el);
-
-    return el;
-  };
-
-  return {htmlInternal, cachedHtmlInternal};
+/** @param {!Array<string>} strings */
+function devAssertCorrectHtmlTemplateTagUsage({length}) {
+  devAssert(length === 1, 'Improper html template tag usage.');
+}
+
+
+/** */
+export class Html {
+
+  /** @param {!./service/ampdoc-impl.AmpDoc} ampdoc */
+  constructor(ampdoc) {
+
+    /** @private @const {!Element} */
+    this.container_ = ampdoc.getRootNode().createElement('div');
+
+    /** @type {?TempCache<!Element>} */
+    this.cache_ = null;
+
+    /**
+     * @param {!Array<string>} strings
+     * @return {!Element}
+     */
+    this.cachedHtmlInternal = strings => { // must be arrow func for scope
```

Comment:
These can be regular methods on the class.

Reply:
Tried i. By passing e.g. `getServiceForDoc(element, SERVICE).htmlInternal` method becomes unbound and breaks.